File: src/main/java/com/micabytes/ink/InkParser.java
package com.micabytes.ink;

import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.Nullable;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.regex.Pattern;

@SuppressWarnings("UtilityClass")
public final class InkParser {
  @NonNls private static final String UTF_8 = "UTF-8";
  @NonNls private static final char WHITESPACE = ' ';
  @NonNls static final char HEADER = '=';
  @NonNls private static final char GATHER_DASH = '-';
  @NonNls static final char CHOICE_DOT = '*';
  @NonNls static final char CHOICE_PLUS = '+';
  @NonNls private static final char VAR_DECL = 'V';
  @NonNls private static final char VAR_STAT = '~';
  @NonNls private static final char CONDITIONAL_HEADER = '{';
  @NonNls static final String CONDITIONAL_END = "}";
  @NonNls public static final char DOT = '.';
  @NonNls private static final String DEFAULT_KNOT_NAME = "default";
  private static final Pattern AT_SPLITTER = Pattern.compile("[@]");
  @NonNls private static final String INCLUDE = "INCLUDE";
  @NonNls private static final String IMG = "img(";

  private InkParser() {
    // NOOP
  }

  public static Story parse(String fileName, StoryWrapper provider) throws InkParseException {
    InputStream in = provider.getStream(fileName);
    Story story = parse(in, provider);
    story.fileName = fileName;
    return story;
  }

  @SuppressWarnings("OverlyComplexMethod")
  public static Story parse(InputStream inputStream, StoryWrapper provider) throws InkParseException {
    if (provider == null)
      throw new InkParseException("The StoryWrapper passed to Ink is null.");
    InputStreamReader inputStreamReader = null;
    BufferedReader bufferedReader = null;
    Story story = new Story(provider);
    try {
      inputStreamReader = new InputStreamReader(inputStream, UTF_8);
      bufferedReader = new BufferedReader(inputStreamReader);
      String line = bufferedReader.readLine();
      int lineNumber = 1;
      Container current = new Knot(lineNumber, "=== " + DEFAULT_KNOT_NAME);
      story.add(current);
      @Nullable Conditional conditional = null;
      while (line != null) {
        String trimmedLine = line.trim();
        if (trimmedLine.contains("//")) {
          String comment = trimmedLine.substring(trimmedLine.indexOf("//")).trim();
          parseComment(lineNumber, comment, current);
          trimmedLine = trimmedLine.substring(0, trimmedLine.indexOf("//")).trim();
        }
        if (trimmedLine.startsWith(INCLUDE)) {
          String file = trimmedLine.replace(INCLUDE, "").trim();
          Story incl = parse(file, provider);
          story.addAll(incl);
        }
        else if (conditional != null) {
          Conditional cond = (Conditional) current.getContent(current.getContentSize()-1);
          cond.parseLine(lineNumber, trimmedLine);
          if (trimmedLine.endsWith(CONDITIONAL_END)) // This is a bug. Means conditions cannot have text line that ends with CBRACE_RIGHT
            conditional = null;
        }
        else {
          Content cont = parseLine(lineNumber, trimmedLine, current);
          if (cont != null) {
            if (cont.getId() == null)
              cont.generateId(current);
            story.add(cont);
            if (cont.isContainer()) {
              current = (Container) cont;
            }
          }
          if (cont != null && cont.isConditional()) {
            conditional = (Conditional) cont;
          }
        }
        line = bufferedReader.readLine();
        lineNumber++;
      }
      story.initialize();
    } catch (IOException e) {
      provider.logException(e);
    } finally {
      try {
        if (inputStreamReader != null)
          inputStreamReader.close();
      } catch (IOException e) {
        provider.logException(e);
      }
      try {
        if (bufferedReader != null)
          bufferedReader.close();
      } catch (IOException e) {
        provider.logException(e);
      }
    }
    return story;
  }

  @SuppressWarnings("OverlyComplexMethod")
  static Content parseLine(int lineNumber, String trimmedLine, Container current) throws InkParseException {
    char firstChar = trimmedLine.isEmpty() ? WHITESPACE : trimmedLine.charAt(0);
    switch (firstChar) {
      case HEADER:
        if (KnotFunction.isFunctionHeader(trimmedLine)) {
          return new KnotFunction(lineNumber, trimmedLine);
        }
        if (Knot.isKnotHeader(trimmedLine)) {
          return new Knot(lineNumber, trimmedLine);
        }
        if (Stitch.isStitchHeader(trimmedLine)) {
          return new Stitch(lineNumber, trimmedLine, current);
        }
        break;
      case CHOICE_DOT:
      case CHOICE_PLUS:
        if (Choice.isChoiceHeader(trimmedLine))
          return new Choice(lineNumber, trimmedLine, current);
        break;
      case GATHER_DASH:
        if (Gather.isGatherHeader(trimmedLine))
          return new Gather(lineNumber, trimmedLine, current);
        break;
      case VAR_DECL:
      case VAR_STAT:
        if (Variable.isVariableHeader(trimmedLine))
          return new Variable(lineNumber, trimmedLine, current);
        break;
      case CONDITIONAL_HEADER:
        if (Conditional.isConditionalHeader(trimmedLine))
          return new Conditional(lineNumber, trimmedLine, current);
        break;
      default:
        break;
    }
    if (!trimmedLine.isEmpty()) {
      return new Content(lineNumber, trimmedLine, current);
    }
    return null;
  }

  private static void parseComment(int lineNumber, String comment, Container current) {
    String[] token = AT_SPLITTER.split(comment);
    if (token.length < 2) return;
    for (int i=1; i<token.length; i++) {
      if (token[i].startsWith(IMG)) {
        String img = token[i].substring(token[i].indexOf(Symbol.BRACE_LEFT) + 1, token[i].indexOf(Symbol.BRACE_RIGHT)).trim();
        if (current != null) {
          current.setBackground(img);
        }
      }
      if (token[i].startsWith("*") || token[i].startsWith("+") ) {
        Content cont = new Comment(lineNumber, token[i]);
        if (current != null)
          current.add(cont);
      }
    }
  }

}

File: src/main/kotlin/com/micabytes/ink/InkParser.kt
package com.micabytes.ink

import com.micabytes.ink.exception.InkParseException
import java.io.InputStream
import java.util.*

object InkParser {

  @Throws(InkParseException::class)
  fun parse(provider: StoryWrapper, fileName: String): Story {
    val input: InputStream = provider.getStream(fileName)
    return parse(input, provider, fileName)
  }

  @Throws(InkParseException::class)
  fun parse(inputStream: InputStream, provider: StoryWrapper, fileName: String): Story {
    val includes = ArrayList<String>()
    val content = HashMap<String, Content>()
    var topContainer: Container? = null
    inputStream.reader(Charsets.UTF_8).buffered(DEFAULT_BUFFER_SIZE).use {
      var line: String? = it.readLine()
      var lineNumber = 1
      var currentContainer: Container? = null
      while (line != null) {
        var trimmedLine = line.trim { it <= ' ' }
        if (trimmedLine.contains(Symbol.COMMENT)) {
          trimmedLine = trimmedLine.substring(0, trimmedLine.indexOf(Symbol.COMMENT)).trim({ it <= ' ' })
        }
        // Tags don't seem to work-
        if (trimmedLine.contains(Symbol.HASHMARK)) {
          val tags = trimmedLine.substring(trimmedLine.indexOf(Symbol.HASHMARK)).trim({ it <= ' ' }).split(Symbol.HASHMARK)
          for (tag in tags) {
            val current = Tag(tag, currentContainer, lineNumber)
            if (currentContainer != null)
              currentContainer.add(current)
            if (!content.containsKey(current.id))
              content.put(current.id, current)
          }
          trimmedLine = trimmedLine.substring(0, trimmedLine.indexOf(Symbol.HASHMARK)).trim({ it <= ' ' })
        }
        if (trimmedLine.startsWith(Symbol.INCLUDE)) {
          val includeFile = trimmedLine.replace(Symbol.INCLUDE, "").trim({ it <= ' ' })
          includes.add(includeFile)
        }
        val tokens = parseLine(lineNumber, trimmedLine, currentContainer)
        for (current in tokens) {
          if (current is Container)
            currentContainer = current
          if (!content.containsKey(current.id))
            content.put(current.id, current)
          if (currentContainer != null && topContainer == null)
            topContainer = currentContainer
        }
        line = it.readLine()
        lineNumber++
      }
    }
    if (topContainer == null) throw InkParseException("Could not detect a root knot node in " + fileName)
    val story = Story(provider, fileName, topContainer!!, content)
    for (includeFile in includes) {
      story.add(parse(provider, includeFile))
    }
    return story
  }

  @Throws(InkParseException::class)
  internal fun parseLine(lineNumber: Int, line: String, currentContainer: Container?): List<Content> {
    val firstChar = if (line.isEmpty()) Symbol.WHITESPACE else line[0]
    when (firstChar) {
      Symbol.HEADER -> {
        if (Knot.isKnot(line)) {
          return mutableListOf(Knot(line, lineNumber))
        }
        if (Stitch.isStitchHeader(line)) {
          if (currentContainer == null)
            throw InkParseException("Stitch without a containing Knot at line " + lineNumber) // add fileName
          return parseContainer(Stitch(line, Stitch.getParent(currentContainer), lineNumber))
        }
      }
      Symbol.CHOICE_DOT, Symbol.CHOICE_PLUS -> {
        if (currentContainer == null)
          throw InkParseException("Choice without an anchor at line " + lineNumber) // add fileName
        val choiceDepth = Choice.getChoiceDepth(line)
        return parseContainer(Choice(line, choiceDepth, Choice.getParent(currentContainer, choiceDepth), lineNumber))
      }
      Symbol.DASH -> {
        if (line.startsWith(Symbol.DIVERT))
          return parseDivert(lineNumber, line, currentContainer)
        if (currentContainer == null)
          throw InkParseException("Dash without an anchor at line " + lineNumber) // add fileName
        if (isConditional(currentContainer)) {
          return parseContainer(ConditionalOption(line, getConditional(currentContainer), lineNumber))
        }
        else {
          val level = Gather.getChoiceDepth(line)
          return parseContainer(Gather(line, Gather.getParent(currentContainer, level), level, lineNumber))
        }
      }
      Symbol.VAR_DECL, Symbol.VAR_STAT -> {
        if (currentContainer == null)
          throw InkParseException("Declaration is not inside a knot/container at line " + lineNumber) // add fileName
        if (Declaration.isVariableHeader(line)) {
          return parseContainer(Declaration(lineNumber, line, currentContainer))
        }
      }
      Symbol.CBRACE_LEFT -> if (Conditional.isConditionalHeader(line))
        return parseContainer(Conditional(line, currentContainer!!, lineNumber))
      Symbol.CBRACE_RIGHT -> if (currentContainer is Conditional || currentContainer is ConditionalOption)
        return mutableListOf(getConditional(currentContainer).parent as Content)
      else -> {
        // NOOP
      }
    }
    if (line.contains(Symbol.DIVERT))
      return parseDivert(lineNumber, line, currentContainer)
    if (!line.isEmpty() && currentContainer != null)
      return parseContainer(Content(Content.getId(currentContainer), line, currentContainer, lineNumber))
    // Should throw error.
    return ArrayList<Content>()
  }

  private fun parseContainer(cont: Content): MutableList<Content> {
    cont.parent!!.add(cont)
    val list = mutableListOf(cont)
    if (cont is Container)
      list.addAll(cont.children)
    return list
  }

  fun isConditional(currentContainer: Container?): Boolean {
    var container = currentContainer
    while (container != null) {
      if (container is Conditional) return true
      container = container.parent
    }
    return false
  }

  fun getConditional(currentContainer: Container): Container {
    var container = currentContainer
    while (container != null) {
      if (container is Conditional) return container
      container = container.parent!!
    }
    return currentContainer
  }

  internal fun parseDivert(lineNumber: Int, line: String, currentContainer: Container?): List<Content> {
    val ret = ArrayList<Content>()
    val div = line.split(Symbol.DIVERT)
    if (!div[0].isEmpty()) {
      val text = Content(Content.getId(currentContainer!!), div[0] + Symbol.GLUE, currentContainer, lineNumber)
      currentContainer.add(text)
      ret.add(text)
    }
    for (i in 1 until div.size) {
      if (div[i].isNotEmpty()) {
        val divert = Divert(lineNumber, div[i], currentContainer!!)
        currentContainer.add(divert)
        ret.add(divert)
      }
    }
    return ret
  }

}
